File: android/src/main/java/com/margelo/nitro/camera/facedetector/HybridFaceDetector.kt
package com.margelo.nitro.camera.facedetector

import android.graphics.Matrix
import androidx.annotation.OptIn
import androidx.camera.core.ExperimentalGetImage
import com.google.mlkit.vision.face.FaceDetection
import com.margelo.nitro.camera.HybridFrameSpec
import com.margelo.nitro.camera.Point as CameraPoint
import com.google.android.gms.tasks.Tasks
import com.margelo.nitro.camera.facedetector.extensions.toMLFaceDetectorOptions
import com.margelo.nitro.camera.facedetector.extensions.toInputImage

class HybridFaceDetector(
  options: FaceDetectorOptions
) : HybridFaceDetectorSpec() {
  private val runLandmarks = options.runLandmarks ?: false
  private val runContours = options.runContours ?: false
  private val runClassifications = options.runClassifications ?: false
  private val trackingEnabled = options.trackingEnabled ?: false
  private val autoMode = options.autoMode ?: false
  private val faceDetector = FaceDetection.getClient(
    options.toMLFaceDetectorOptions()
  )

  @OptIn(ExperimentalGetImage::class)
  override fun detectFaces(
    frame: HybridFrameSpec
  ): Array<HybridFaceSpec> {
    val image = frame.toInputImage()
    val (mlKitFrameWidth, mlKitFrameHeight) = getMLKitCoordinateDimensions(
      image.width.toDouble(),
      image.height.toDouble(),
      image.rotationDegrees
    )
    // ML Kit returns points after applying InputImage.rotationDegrees. Undo that
    // rotation before sampling VisionCamera's native raw-frame -> camera transform.
    val pointTransformer = if (autoMode) {
      createMLKitToCameraPointTransformer(
        frame,
        frame.width.toInt(),
        frame.height.toInt(),
        image.rotationDegrees
      )
    } else createIdentityPointTransformer()

    val config = createFaceProcessConfig(
      frameWidth = mlKitFrameWidth,
      frameHeight = mlKitFrameHeight,
      autoMode = autoMode,
      runLandmarks = runLandmarks,
      runContours = runContours,
      runClassifications = runClassifications,
      trackingEnabled = trackingEnabled,
      pointTransformer = pointTransformer
    )
    val task = faceDetector.process(image)
    val faces = Tasks.await(task).map {
      HybridFace(it, config)
    }.toTypedArray<HybridFaceSpec>()

    return faces
  }

  override fun dispose() {
    faceDetector.close()
    super.dispose()
  }

  private fun createMLKitToCameraPointTransformer(
    frame: HybridFrameSpec,
    frameWidth: Int,
    frameHeight: Int,
    rotationDegrees: Int
  ): (Double, Double) -> Pair<Double, Double> {
    val mlKitToFrame = Matrix()
    check(
      createBufferToMLKitRotationMatrix(
        frameWidth,
        frameHeight,
        rotationDegrees
      ).invert(mlKitToFrame)
    ) {
      "Could not invert ML Kit buffer rotation matrix."
    }

    fun convertPoint(x: Double, y: Double): CameraPoint {
      val point = floatArrayOf(x.toFloat(), y.toFloat())
      mlKitToFrame.mapPoints(point)
      return frame.convertFramePointToCameraPoint(
        CameraPoint(point[0].toDouble(), point[1].toDouble())
      )
    }

    // Capture the composed affine transform while the Frame is still valid.
    val origin = convertPoint(0.0, 0.0)
    val xAxis = convertPoint(1.0, 0.0)
    val yAxis = convertPoint(0.0, 1.0)

    return { 
      x, y -> Pair(
        origin.x + (xAxis.x - origin.x) * x + (yAxis.x - origin.x) * y,
        origin.y + (xAxis.y - origin.y) * x + (yAxis.y - origin.y) * y
      )
    }
  }
}

File: ios/HybridFaceDetector.swift
import CoreImage
import NitroModules
import Vision
import VisionCamera

private typealias VisionCameraPoint = margelo.nitro.camera.Point

class HybridFaceDetector: HybridFaceDetectorSpec {
  private let runLandmarks: Bool
  private let runContours: Bool
  private let runClassifications: Bool
  private let trackingEnabled: Bool
  private let autoMode: Bool

  init(_ options: FaceDetectorOptions) {
    self.runLandmarks = options.runLandmarks ?? false
    self.runContours = options.runContours ?? false
    self.runClassifications = options.runClassifications ?? false
    self.trackingEnabled = options.trackingEnabled ?? false
    self.autoMode = options.autoMode ?? false
    super.init()
  }

  func detectFaces(frame: any HybridFrameSpec) throws -> [any HybridFaceSpec] {
    guard let nativeFrame = frame as? any NativeFrame else {
      throw RuntimeError.error(withMessage: "Frame is not of type `NativeFrame`!")
    }
    guard let sampleBuffer = nativeFrame.sampleBuffer,
          let pixelBuffer = CMSampleBufferGetImageBuffer(sampleBuffer) else {
      throw RuntimeError.error(withMessage: "Frame doesn't contain a valid CVPixelBuffer!")
    }

    let request = VNDetectFaceLandmarksRequest()
    let handler = VNImageRequestHandler(
      cvPixelBuffer: pixelBuffer,
      orientation: frame.orientation.toCGImagePropertyOrientation(isMirrored: frame.isMirrored),
      options: [:]
    )
    try handler.perform([request])

    let config = createFaceProcessConfig(
      frame.width,
      frame.height,
      autoMode,
      runLandmarks,
      runContours,
      runClassifications,
      trackingEnabled,
      try createFrameToCameraPointTransformer(frame)
    )

    return (request.results ?? []).map { HybridFace(face: $0, config: config) }
  }

  private func createFrameToCameraPointTransformer(
    _ frame: any HybridFrameSpec
  ) throws -> (Double, Double) -> Point {
    let origin = try frame.convertFramePointToCameraPoint(framePoint: VisionCameraPoint(0.0, 0.0))
    let xAxis = try frame.convertFramePointToCameraPoint(framePoint: VisionCameraPoint(1.0, 0.0))
    let yAxis = try frame.convertFramePointToCameraPoint(framePoint: VisionCameraPoint(0.0, 1.0))
    let originX = origin.x
    let originY = origin.y
    let xAxisDeltaX = xAxis.x - originX
    let xAxisDeltaY = xAxis.y - originY
    let yAxisDeltaX = yAxis.x - originX
    let yAxisDeltaY = yAxis.y - originY

    return { x, y in
      Point(
        x: originX + xAxisDeltaX * x + yAxisDeltaX * y,
        y: originY + xAxisDeltaY * x + yAxisDeltaY * y
      )
    }
  }
}
